File: frameworks/playwright-ts/pages/LoginPage.ts
import { Page } from "@playwright/test";
import { BasePage } from "./BasePage";

export class LoginPage extends BasePage {
  readonly usernameInput = this.page.locator("#username");
  readonly passwordInput = this.page.locator("#password");
  readonly submitButton = this.page.locator("#login-form button[type=submit]");
  readonly errorBox = this.page.locator("#login-error");

  constructor(page: Page) {
    super(page);
  }

  async open() {
    await this.goto("/login");
  }

  async login(username: string, password: string) {
    await this.usernameInput.fill(username);
    await this.passwordInput.fill(password);
    await this.submitButton.click();
  }
}

File: frameworks/playwright-dotnet/Pages/LoginPage.cs
using Microsoft.Playwright;

namespace QARonin.PlaywrightDotnet.Pages;

public class LoginPage
{
    private readonly IPage _page;
    public LoginPage(IPage page) => _page = page;

    public ILocator UsernameInput => _page.Locator("#username");
    public ILocator PasswordInput => _page.Locator("#password");
    public ILocator SubmitButton => _page.Locator("button[type=submit]");
    public ILocator ErrorBox => _page.Locator(".error, #error");

    public async Task OpenAsync(string baseUrl) => await _page.GotoAsync($"{baseUrl}/login");

    public async Task LoginAsync(string username, string password)
    {
        await UsernameInput.FillAsync(username);
        await PasswordInput.FillAsync(password);
        await SubmitButton.ClickAsync();
    }
}
